File: tests/test_unnamed_namespace_a.py
from __future__ import annotations

import pytest

from pybind11_tests import unnamed_namespace_a as m
from pybind11_tests import unnamed_namespace_b as mb

XFAIL_CONDITION = "not m.defined_WIN32_or__WIN32 and (m.defined___clang__ or m.defined__LIBCPP_VERSION)"
XFAIL_REASON = "Known issues: https://github.com/pybind/pybind11/pull/4319"


@pytest.mark.xfail(XFAIL_CONDITION, reason=XFAIL_REASON, strict=False)
@pytest.mark.parametrize(
    "any_struct", [m.unnamed_namespace_a_any_struct, mb.unnamed_namespace_b_any_struct]
)
def test_have_class_any_struct(any_struct):
    assert any_struct is not None


def test_have_at_least_one_class_any_struct():
    assert (
        m.unnamed_namespace_a_any_struct is not None
        or mb.unnamed_namespace_b_any_struct is not None
    )


@pytest.mark.xfail(XFAIL_CONDITION, reason=XFAIL_REASON, strict=True)
def test_have_both_class_any_struct():
    assert m.unnamed_namespace_a_any_struct is not None
    assert mb.unnamed_namespace_b_any_struct is not None

File: tests/test_unnamed_namespace_a.cpp
#include "pybind11_tests.h"

namespace {
struct any_struct {};
} // namespace

TEST_SUBMODULE(unnamed_namespace_a, m) {
    if (py::detail::get_type_info(typeid(any_struct)) == nullptr) {
        py::class_<any_struct>(m, "unnamed_namespace_a_any_struct");
    } else {
        m.attr("unnamed_namespace_a_any_struct") = py::none();
    }
    m.attr("defined___clang__") =
#if defined(__clang__)
        true;
#else
        false;
#endif
    m.attr("defined__LIBCPP_VERSION") =
#if defined(_LIBCPP_VERSION)
        true;
#else
        false;
#endif
}
